src/coreclr/src/jit/earlyprop.cpp

```diff
@@ -83,17 +85,32 @@ GenTree* Compiler::getArrayLengthFromAllocation(GenTree* tree DEBUGARG(BasicBloc
                 call->gtCallMethHnd == eeFindHelper(CORINFO_HELP_NEWARR_1_VC) ||
                 call->gtCallMethHnd == eeFindHelper(CORINFO_HELP_NEWARR_1_ALIGN8))
             {
-#ifdef DEBUG
-                optCheckFlagsAreSet(OMF_HAS_NEWARRAY, "OMF_HAS_NEWARRAY", BBF_HAS_NEWARRAY, "BBF_HAS_NEWARRAY", tree,
-                                    block);
-#endif
                 // This is an array allocation site. Grab the array length node.
-                return gtArgEntryByArgNum(call, 1)->GetNode();
+                rtn = gtArgEntryByArgNum(call, 1)->GetNode();
+            }
+            else if (call->gtCallMethHnd == eeFindHelper(CORINFO_HELP_READYTORUN_NEWARR_1))
+            {
+#ifdef TARGET_ARMARCH
+                // On arm, Ready To Run will, on some platforms, insert a handle before the length.
+                // In this case, we grab the last argument, as that will always be the length.
+                // See fgInitArgInfo
+                int arrLenArgNum = call->fgArgInfo->ArgCount() - 1;
+#else
+                int arrLenArgNum = 0;
+#endif
+                rtn = gtArgEntryByArgNum(call, arrLenArgNum)->GetNode();
```

Comment:
Interesting, this could explain an early return in this block:
https://github.com/dotnet/runtime/blob/7360c191971156855a07b93c97f70affc622c37f/src/coreclr/jit/importer.cpp#L3511-L3529

Could you please try to extract this logic with an assert from that block (`assert(arrayLengthNode->OperIs(GT_CNS_INT))` and use it in both places?

Is it correct to always use the last on both arm and not-arm for R2R?

Reply:
Yes, it's always correct to use the last. I've changed it over to use that unconditionally. 

I'm not following what you mean with that assert. At this point, we can get any arbitrary node as an argument. That optimization only works when the size is constant.

I don't think this can be easily combined. fgArgInfo isn't built yet in the importer, and gtCallArgs's nodes are changed to fgArgPlace nodes somewhere between the importer and earlyprop.

Reply:
> I'm not following what you mean with that assert. At this point, we can get any arbitrary node as an argument. That optimization only works when the size is constant.

If I understand correctly this code would get an incorrect node on arm R2R, it takes the first instead of the second, but the next check prevents the optimization so we don't generate anything bad.

Reply:
got it, thank you for the explanation!